raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotWriter.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.snapshot;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.nio.file.StandardOpenOption;
+import org.apache.kafka.common.utils.Utils;
+import org.apache.kafka.raft.OffsetAndEpoch;
+
+public final class FileRawSnapshotWriter implements RawSnapshotWriter {
+    private final Path tempSnapshotPath;
+    private final FileChannel channel;
+    private final OffsetAndEpoch snapshotId;
+    private boolean frozen = false;
+
+    private FileRawSnapshotWriter(
+        Path tempSnapshotPath,
+        FileChannel channel,
+        OffsetAndEpoch snapshotId
+    ) {
+        this.tempSnapshotPath = tempSnapshotPath;
+        this.channel = channel;
+        this.snapshotId = snapshotId;
+    }
+
+    @Override
+    public OffsetAndEpoch snapshotId() {
+        return snapshotId;
+    }
+
+    @Override
+    public long sizeInBytes() throws IOException {
+        return channel.size();
+    }
+
+    @Override
+    public void append(ByteBuffer buffer) throws IOException {
+        if (frozen) {
+            throw new IllegalStateException(
+                String.format("Append not supported. Snapshot is already frozen: id = %s; temp path = %s", snapshotId, tempSnapshotPath)
+            );
+        }
+
+        Utils.writeFully(channel, buffer);
+    }
+
+    @Override
+    public boolean isFrozen() {
+        return frozen;
+    }
+
+    @Override
+    public void freeze() throws IOException {
+        channel.close();
+        frozen = true;
+
+        // Set readonly and ignore the result
+        if (!tempSnapshotPath.toFile().setReadOnly()) {
+            throw new IOException(String.format("Unable to set file (%s) as read-only", tempSnapshotPath));
+        }
+
+        Path destination = Snapshots.moveRename(tempSnapshotPath, snapshotId);
+        Files.move(tempSnapshotPath, destination, StandardCopyOption.ATOMIC_MOVE);
```

Comment:
It might be useful to review the history behind `Utils.atomicMoveWithFallback`. It's not clear to me why this case is different from some of the other situations that it is used.

Reply:
I investigated the issue a bit. It look like this utility was introduce to replace these lines in `OffsetCheckpoint.scala`:
```scala
// swap new offset checkpoint file with previous one
if(!temp.renameTo(file)) {
    // renameTo() fails on Windows if the destination file exists.
    file.delete()
    if(!temp.renameTo(file))
        throw new IOException(...)
}
```

Looking at the JDK implementation in Windows, `ATOMIC_MOVE` should work in Windows if the target file exists: https://github.com/openjdk/jdk/blob/master/src/java.base/windows/classes/sun/nio/fs/WindowsFileCopy.java#L298-L312

In other words, I think we can keep the code as is in this PR.

cc @ijuma

Reply:
It's not only Windows, NFS also has some restrictions. Why don't you want to use the utility method?

Reply:
Also, I would not rely on reading the code to assume one way or another. You'd want to test it too.